### ISSUE TITLE
Not clear how to reposition seaborn.histplot legend

### ISSUE THREAD

**Issue by mortonjt** [mwaskom/seaborn#2280]
I was pretty excited about seeing the `histplot` functionality coming out, and I'm getting *beautiful* figures with it already.

I'm trying to finetune my figures and reposition the legends - but am running into issues when moving the legend.
Below is toy example 
```python
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
fig, ax = plt.subplots()
x = np.hstack((np.random.normal(0, 1, 100), np.random.normal(-0.5, 1, 100), np.random.normal(0.5, 1, 100)))
data = pd.DataFrame({'x': x, 'd' : ['a'] * 100 + ['b'] * 100 + ['c'] * 100})
g = sns.histplot(data, x='x', hue="d", element="step", stat="probability", ax=ax)
```
![image](https://user-images.githubusercontent.com/4184797/93139918-d1012400-f69e-11ea-9bc8-74b4779adb06.png)

When I move the legend to the upper left, the legend disappears with the warning `No handles with labels found to put in legend.`.
```python
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
fig, ax = plt.subplots()
x = np.hstack((np.random.normal(0, 1, 100), np.random.normal(-0.5, 1, 100), np.random.normal(0.5, 1, 100)))
data = pd.DataFrame({'x': x, 'd' : ['a'] * 100 + ['b'] * 100 + ['c'] * 100})
g = sns.histplot(data, x='x', hue="d", element="step", stat="probability", ax=ax, legend=False)
ax.legend(loc='upper-left')
```

![image](https://user-images.githubusercontent.com/4184797/93140027-fe4dd200-f69e-11ea-8135-ffd85a627d30.png)

This is interesting, because it doesn't appear to be an issue with the other plots
https://stackoverflow.com/questions/27019079/move-seaborn-plot-legend-to-a-different-position
https://stackoverflow.com/questions/53733755/how-to-move-legend-to-outside-of-a-seaborn-scatterplot/53737271

I'm curious if this is unique to `histplot` (or if I'm overlooking something).

**Comment by mwaskom** [comment]
Hi, yes, legends do work a little bit differently in the distribution plots, and they're also still a little rough around the edges.

In short, seaborn currently has two options: add a bunch of phantom artists to the axes with labels (meaning that calling `ax.legend()` again will pick them up and replace the existing legend) or pass handles and labels directly to `ax.legend` (meaning that inspecting or modifying the artists is easier, but "moving" the legend is harder).

IMO the root problem is that matplotlib legend objects lack a public-facing API for modifying them once they exist. I'm pushing on matplotlib to fix this limitation in core, which would be better than implementing a hacky work around in seaborn ([thread here](https://discourse.matplotlib.org/t/legends-are-difficult-to-modify/21506)).

Here's also a issue tracking some known ongoing difficulties with legends in seaborn, some (but not all) of which are attributable to the matplotlib-level API: https://github.com/mwaskom/seaborn/issues/2231. Solving some of these will require moving way from "option 1" above.

Right now, your best bet is probably something like this:

```python

def move_legend(ax, new_loc, **kws):
    old_legend = ax.legend_
    handles = old_legend.legendHandles
    labels = [t.get_text() for t in old_legend.get_texts()]
    title = old_legend.get_title().get_text()
    ax.legend(handles, labels, loc=new_loc, title=title, **kws)
    
move_legend(ax, "upper left")
```

**Comment by felipezeiser** [comment]
> Hi, yes, legends do work a little bit differently in the distribution plots, and they're also still a little rough around the edges.
> 
> In short, seaborn currently has two options: add a bunch of phantom artists to the axes with labels (meaning that calling `ax.legend()` again will pick them up and replace the existing legend) or pass handles and labels directly to `ax.legend` (meaning that inspecting or modifying the artists is easier, but "moving" the legend is harder).
> 
> IMO the root problem is that matplotlib legend objects lack a public-facing API for modifying them once they exist. I'm pushing on matplotlib to fix this limitation in core, which would be better than implementing a hacky work around in seaborn ([thread here](https://discourse.matplotlib.org/t/legends-are-difficult-to-modify/21506)).
> 
> Here's also a issue tracking some known ongoing difficulties with legends in seaborn, some (but not all) of which are attributable to the matplotlib-level API: #2231. Solving some of these will require moving way from "option 1" above.
> 
> Right now, your best bet is probably something like this:
> 
> ```python
> def move_legend(ax, new_loc, **kws):
>     old_legend = ax.legend_
>     handles = old_legend.legendHandles
>     labels = [t.get_text() for t in old_legend.get_texts()]
>     title = old_legend.get_title().get_text()
>     ax.legend(handles, labels, loc=new_loc, title=title, **kws)
>     
> move_legend(ax, "upper left")
> ```

this worked for me. thank you so much

**Comment by mwaskom** [comment]
I'm going to consider this closed with #2643, which codifies the `move_legend` hack as a helper function in seaborn. It doesn't solve the "`histplot` works differently from other plots" problem, but I think that's unlikely to change in the future (or rather, for reasons discussed above, other functions will work more like `histplot`).

So, official recommendation is to move (and update) the legend, for any seaborn plot, like this:

```python
ax = sns.histplot(data=penguins, x="bill_length_mm", hue="species")
sns.move_legend(ax, "lower center", bbox_to_anchor=(.5, 1), ncol=3, title_fontsize=14)
```
![image](https://user-images.githubusercontent.com/315810/129490722-341a3d7d-8242-4232-8c60-f9ca730b41c5.png)

**Comment by erinboyle** [comment]
Thank you!! Big improvement!

**Comment by jofa974** [comment]
I was about to go crazy... until I found this post. Thanks a lot !

**Comment by ravshanovbek** [comment]
> I'm going to consider this closed with #2643, which codifies the `move_legend` hack as a helper function in seaborn. It doesn't solve the "`histplot` works differently from other plots" problem, but I think that's unlikely to change in the future (or rather, for reasons discussed above, other functions will work more like `histplot`). Legend inconsistencies / difficulties are otherwise tracked in #2231.
> 
> So, official recommendation is to move (and update) the legend, for any seaborn plot, like this:
> 
> ```python
> ax = sns.histplot(data=penguins, x="bill_length_mm", hue="species")
> sns.move_legend(ax, "lower center", bbox_to_anchor=(.5, 1), ncol=3, title_fontsize=14)
> ```
> 
> ![image](https://user-images.githubusercontent.com/315810/129490722-341a3d7d-8242-4232-8c60-f9ca730b41c5.png)
Thank you!
